sklearn/preprocessing/tests/test_target_encoder.py

```diff
@@ -0,0 +1,304 @@
+        cur_encodings = _encode_target(X_, y_, n_categories, smooth)
+        expected_X_fit_transform[test_idx, 0] = cur_encodings[X_int[test_idx, 0]]
+
+    target_encoder = TargetEncoder(smooth=smooth, categories=categories, cv=kfold)
+
+    X_fit_transform = target_encoder.fit_transform(X_input, y_input)
+
+    assert target_encoder.target_type_ == target_type
+    assert_allclose(X_fit_transform, expected_X_fit_transform)
+    assert len(target_encoder.encodings_) == 1
+    assert_allclose(target_encoder.encodings_[0], expected_encodings)
+    assert target_encoder.encoding_mean_ == pytest.approx(y_mean)
+
+    X_test = np.array([[0, 1, 2]], dtype=np.int64).T
+    expected_X_test_transform = np.concatenate(
+        (expected_encodings, np.asarray([y_mean]))
+    ).reshape(-1, 1)
+
+    if isinstance(categories, str) and categories == "auto":
+        X_test_input = X_test
+    else:
+        X_test_input = categories[0][X_test]
+
+    X_test_input = np.concatenate((X_test_input, [[unknown_value]]))
+
+    X_test_transform = target_encoder.transform(X_test_input)
+    assert_allclose(X_test_transform, expected_X_test_transform)
+
+
+@pytest.mark.parametrize(
+    "X, categories",
+    [
+        (
+            np.array([[0] * 10 + [1] * 10 + [3]], dtype=np.int64).T,  # 3 is unknown
+            [[0, 1, 2]],
+        ),
+        (
+            np.array(
+                [["cat"] * 10 + ["dog"] * 10 + ["snake"]], dtype=object
+            ).T,  # snake is unknown
+            [["dog", "cat", "cow"]],
+        ),
+    ],
+)
+@pytest.mark.parametrize("smooth", [4.0, "auto"])
+def test_custom_categories(X, categories, smooth):
+    """custom categoires with unknown categories that are not in training data."""
+    rng = np.random.RandomState(42)
+    y = rng.uniform(low=-10, high=20, size=X.shape[0])
+    enc = TargetEncoder(categories=categories, smooth=smooth).fit(X, y)
+
+    # The last element is unknown and encoded as the mean
+    y_mean = y.mean()
+    X_trans = enc.transform(X[-1:])
+    assert X_trans[0, 0] == pytest.approx(y_mean)
+
+    assert len(enc.encodings_) == 1
+    # custom category that is not in training data
+    assert enc.encodings_[0][-1] == pytest.approx(y_mean)
+
+
+@pytest.mark.parametrize(
+    "y, msg",
+    [
+        ([1, 2, 0, 1], "Found input variables with inconsistent"),
+        (
+            np.asarray([[1, 2, 0], [1, 2, 3]]).T,
+            "Target type was inferred to be 'multiclass-multioutput'",
+        ),
+        (["cat", "dog", "bear"], "Target type was inferred to be 'multiclass'"),
+    ],
+)
+def test_errors(y, msg):
+    """Check invalidate input."""
+    X = np.asarray([[1, 0, 1]]).T
+
+    enc = TargetEncoder()
+    with pytest.raises(ValueError, match=msg):
+        enc.fit_transform(X, y)
+
+
+def test_use_regression_target():
+    """Custom target_type to avoid inferring the target type."""
+    X = np.asarray([[0, 1, 0, 1, 0, 1]]).T
+    # type_of_target would be 'multiclass'
```

Comment:
that seems bad? at least once we support multiclass this would be a silent unexpected behavior, right?

Reply:
I think an assumption of type_of_target is that any model only supports classification or regression, but here we support both?